### PR TITLE
Replace pre-commit checks with ruff

### PR DESCRIPTION
**Pull request by dhimmel**
[Ruff](https://github.com/charliermarsh/ruff#pycodestyle-e-w) is an "extremely fast Python linter, written in Rust". It replaces several of our existing tools including flake8, pyupgrade, and isort. It uses the same error codes when possible. Has been gaining popularity recently with rave reviews, so giving it a try.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -8,3 +8,24 @@ addopts = "--color=yes -rs"
...
+select = [
+    "B",  # flake8-bugbear
+    "C",  # flake8-comprehensions
+    "C90",  # mccabe
+    "E",  # pycodestyle errors
+    "F",  # pyflakes
+    "I",  # isort
+    "UP",  # pyupgrade
+    "W",  # pycode warnings
+]
```
Comment (dhimmel) [inline pyproject.toml:28]: Here is where we set rule codes to enable.

**Comment by dhimmel** [comment]
Will merge now, since want to prevent conflicts with other PRs.

File: manubot/cite/citekey.py [hunk 2]
```diff
@@ -249,13 +249,16 @@ def shorten_citekey(standard_citekey: str) -> str:
 
 
 def citekey_to_csl_item(
-    citekey, prune=True, manual_refs={}, log_level: tp.Union[str, int] = "WARNING"
+    citekey, prune=True, manual_refs=None, log_level: tp.Union[str, int] = "WARNING"
```
Comment (agitter) [inline manubot/cite/citekey.py:252]: This is a good change. I recently was reminded of the unexpected behavior of this pattern.

File: setup.cfg [hunk 3]
```diff
@@ -80,18 +80,3 @@ manubot =
     cite/curie/*.json
     process/header-includes-template.html
     webpage/redirect-template.html
-
-[flake8]
-ignore = E203, E266, E501, F541, W503
```
Comment (agitter) [inline setup.cfg:85]: Are we choosing to modify the flake8 behavior by no longer ignoring these?

File: manubot/cite/tests/test_doi.py [hunk 4]
```diff
@@ -42,7 +42,9 @@ def test_get_doi_csl_item_zotero():
     doi = "10.1038/ng.3834"
     csl_item = get_doi_csl_item_zotero(doi)
     assert isinstance(csl_item, dict)
-    csl_item["author"][9]["family"] == "GTEx Consortium"
+    # for an unknown reason, GTEx Consortium becomes the first author
```
Comment (agitter) [inline manubot/cite/tests/test_doi.py:45]: This is very strange, and I don't understand why this reference has multiple inconsistent versions of the metadata.

**Review by agitter — COMMENTED** [review]
I wanted to understand this pull request so I'm leaving a post-merge review. Everything looks good, especially ruff itself.

File: setup.cfg (hunk 3 above)
Comment (dhimmel) [reply setup.cfg:85]: According to https://lintlyci.github.io/Flake8Rules/, here are what those rules are:

```
E203	Whitespace before ':'
E266	Too many leading '#' for block comment
E501	Line too long (82 > 79 characters)
W503	Line break occurred before a binary operator
```

It's possible that we're no longer ignoring them if ruff implements and enforces them based on our selection. IIRC I just copied these rules from elsewhere, so probably not critical to ignore them.

**Pull request merged**